ghauth.js

```diff
@@ -35,123 +41,174 @@ function newlineify (len, str) {
   return s
 }
 
-async function createAuth (options) {
-  const reqOptions = {
-    headers: {
-      'X-GitHub-OTP': options.otp || null,
-      'User-Agent': options.userAgent || defaultUA,
-      'Content-type': 'application/json'
-    },
-    method: 'post',
-    auth: `${options.user}:${options.pass}`
-  }
-  const authUrl = options.authUrl || defaultAuthUrl
-  const currentDate = new Date().toJSON()
+function delay (s) {
+  const ms = s * 1000
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
 
+async function hyperquestJson (url, query, reqOptions) {
   const jsonData = await new Promise((resolve, reject) => {
-    const req = hyperquest(authUrl, reqOptions)
-
+    const req = hyperquest(`${url}?${querystring.stringify(query)}`, reqOptions)
     req.pipe(bl((err, data) => {
       if (err) {
         return reject(err)
       }
       resolve(data)
     }))
-
-    req.end(JSON.stringify({
-      scopes: options.scopes || defaultScopes,
-      note: `${(options.note || defaultNote)} (${currentDate})`
-    }))
+    req.end()
   })
 
-  const data = JSON.parse(jsonData.toString())
-
-  if (data.message) {
-    const error = new Error(data.message)
-    error.data = data
-    throw error
-  }
-
-  if (!data.token) {
-    throw new Error('No token from GitHub!')
-  }
-
-  return data.token
+  return JSON.parse(jsonData.toString())
 }
 
 // prompt the user for credentials
 async function prompt (options) {
   const promptName = options.promptName || defaultPromptName
+  const deviceCodeUrl = options.deviceCodeUrl || defaultDeviceCodeUrl
   const accessTokenUrl = options.accessTokenUrl || defaultAccessTokenUrl
   const scopes = options.scopes || defaultScopes
-  const usernamePrompt = options.usernamePrompt || `Your ${promptName} username:`
-  const tokenQuestionPrompt = options.tokenQuestionPrompt || 'This appears to be a personal access token, is that correct? [y/n] '
   const passwordReplaceChar = options.passwordReplaceChar || defaultPasswordReplaceChar
-  let passwordPrompt = options.passwordPrompt
+  const spinner = ora()
 
-  if (!passwordPrompt) {
-    let patMsg = `You may either enter your ${promptName} password or use a 40 character personal access token generated at ${accessTokenUrl} ` +
-      (scopes.length ? `with the following scopes: ${scopes.join(', ')}` : '(no scopes necessary)')
-    patMsg = newlineify(80, patMsg)
-    passwordPrompt = `${patMsg}\nYour ${promptName} password:`
+  let endDeviceFlow = false
```

Comment:
Need to comment on this explaining that this is how we communicate promise cancellation.  A cancellable promise might clarify things more, but this works fine.